lib/ret_web/controllers/api/v1/avatar_controller.ex

```diff
@@ -0,0 +1,192 @@
+
+        case result do
+          :ok ->
+            conn |> render("create.json", avatar: avatar)
+
+          :error ->
+            conn |> send_resp(422, "invalid avatar")
+        end
+
+      {:error, :not_found} ->
+        conn |> send_resp(404, "no such file(s)")
+
+      {:error, :not_allowed} ->
+        conn |> send_resp(401, "")
+    end
+  end
+
+  defp collapse_avatar_files(%{parent_avatar: nil} = avatar),
+    do: avatar |> Map.take(@file_columns)
+
+  defp collapse_avatar_files(%{parent_avatar: parent} = avatar) do
+    Map.merge(collapse_avatar_files(parent), avatar |> Map.take(@file_columns), fn _k, v1, v2 ->
+      if is_nil(v2), do: v1, else: v2
+    end)
+  end
+
+  defp get_avatar(avatar_sid) do
+    avatar =
+      Avatar
+      |> Repo.get_by(avatar_sid: avatar_sid)
+      |> Repo.preload([@file_columns ++ [:parent_avatar, :account]])
+  end
+
+  def show(conn, %{"id" => avatar_sid}) do
+    conn |> show(avatar_sid |> get_avatar())
+  end
+
+  def show(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show(conn, %Avatar{} = avatar) do
+    conn |> render("show.json", avatar: avatar)
+  end
+
+  def show_gltf(conn, %{"id" => avatar_sid}) do
+    conn |> show_gltf(Avatar |> Repo.get_by(avatar_sid: avatar_sid))
+  end
+
+  def show_gltf(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show_gltf(conn, %Avatar{} = avatar) do
+    # TODO we ideally don't need to be featching the OwnedFiles until after we collapse them
+    avatar_files =
+      avatar
+      |> Avatar.load_parents(@file_columns)
+      |> Map.from_struct()
+      |> collapse_avatar_files()
+
+    case Storage.fetch(avatar_files.gltf_owned_file) do
+      {:ok, %{"content_type" => content_type, "content_length" => content_length}, stream} ->
+        gltf =
+          stream
+          |> Enum.join("")
+          |> Poison.decode!()
+          |> with_material(@primary_material_name, avatar_files |> Map.take(@image_columns))
+          |> with_buffer(avatar_files.bin_owned_file)
+
+        conn
+        # |> put_resp_content_type("model/gltf", nil)
+        |> send_resp(200, gltf |> Poison.encode!())
+
+      {:error, :not_found} ->
+        conn |> send_resp(404, "")
+
+      {:error, :not_allowed} ->
+        conn |> send_resp(401, "")
+    end
+  end
+
+  defp with_buffer(gltf, bin_file) do
+    gltf
+    |> Map.replace("buffers", [
+      %{
+        uri: bin_file |> OwnedFile.uri_for() |> URI.to_string()
+      }
+    ])
+  end
+
+  defp with_material(gltf, name, image_files) do
+    case gltf["materials"] |> Enum.find(&(&1["name"] == name)) do
+      nil ->
+        gltf
+
+      material ->
+        image_files
+        |> Enum.filter(fn {_, v} -> v end)
+        |> Enum.flat_map(fn {col, file} ->
+          Enum.map(@texture_paths[col], fn path ->
+            texture_index = material |> get_in(path ++ ["index"])
+            image_index = gltf |> get_in(["textures", Access.at(texture_index), "source"])
```

Comment:
i don't usually see `Access` referenced directly, it may be more idiomatic to just use `Enum.at`

Reply:
Hmm, `Enum.at` does not appear to be a function. I am pretty sure this is the "correct" way to do this sort of programmatic access with the `X_in` functions. https://hexdocs.pm/elixir/Access.html#at/1. Looking this up I also just learned about `Access.all` which is pretty nifty.

Reply:
cool. Enum.at def exists fyi https://hexdocs.pm/elixir/Enum.html#at/3

Reply:
cool. Enum.at def exists fyi https://hexdocs.pm/elixir/Enum.html#at/3

Reply:
oh i was mis-reading what this did. didn't realize Access.at was higher order, but clearly that makes sense

Reply:
Sorry, I meant `Enum.at/1` does not exist.